File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/C-Длина_максимального_подпалиндрома/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N^2)
    Memory Complexity O(N^2)
    """
    s: str = sys.stdin.readline().rstrip()
    n: int = len(s)
    dp: list[list[int]] = [[0] * n for _ in range(n)]
    for i in range(n): dp[i][i] = 1
    for length in range(2, n + 1):
        for left in range(n - length + 1):
            right: int = left + length - 1
            if s[left] == s[right]: dp[left][right] = 2 + [dp[left + 1][right - 1], 0][length == 2]
            else: dp[left][right] = max(dp[left + 1][right], dp[left][right - 1])
    length: int = dp[0][n - 1]
    output: list[str] = []
    left: int = 0
    right: int = n - 1
    while left <= right:
        if s[left] == s[right]:
            output.append(s[left])
            left += 1
            right -= 1
        else:
            if dp[left + 1][right] > dp[left][right - 1]: left += 1
            else: right -= 1
    reversed_part: list[str] = output[:len(output) - (length & 1):][::-1]
    output.extend(reversed_part)
    sys.stdout.write('{}\n{}\n'.format(len(output), ''.join(output)))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/C-Длина_максимального_подпалиндрома/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;


void solution() {
    /*
    Time Complexity O(N^2)
    Memory Complexity O(N^2)
    */
    string s;
    cin >> s;
    int n = s.size();
    vector<vector<int>> dp(n, vector<int>(n, 0));
    for (int i = 0; i < n; ++i) dp[i][i] = 1;

    for (int len = 2; len <= n; ++len) {
        for (int i = 0; i <= n - len; ++i) {
            int j = i + len - 1;
            if (s[i] == s[j]) dp[i][j] = 2 + (len == 2 ? 0 : dp[i + 1][j - 1]);
            else dp[i][j] = max(dp[i + 1][j], dp[i][j - 1]);
        }
    }

    int length = dp[0][n-1];
    string output;
    int i = 0, j = n - 1;
    while (i <= j) {
        if (s[i] == s[j]) {
            output += s[i];
            ++i;
            --j;
        } else {
            if (dp[i + 1][j] > dp[i][j - 1]) ++i;
            else --j;
        }
    }
    string reversed_part = output.substr(0, output.size() - (length & 1));
    reverse(reversed_part.begin(), reversed_part.end());
    output += reversed_part;
    printf("%d\n", output.size());
    cout << output << "\n";
}


int main() {
    solution();
}